Raise AR::IrreversibleOrderError when #reverse_order can not do it's job

bogdan opened the pull request:
Current `#reverse_order` algorithm is very simple and doesn't try to
handle complicated cases of reversing order given as string.
This patch makes AR to raise exception instead of generating invalid SQL
inside `#reverse_order`.

Fixes #8225, #11571

This patch is alternative to #16779 and #18754.

sgrif commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -1009,14 +1009,24 @@ def arel_columns(columns)
           o.reverse
         when String
-          o.to_s.split(',').map! do |s|
+          if o =~ /\([^()]*,[^()]*\)/
```
Can we pull this out into a method that states what it's looking for, like `contains_functions_or_complex_sql_expressions_or_whatever_i_dont_thing_regexes_are_great_at_expressing_intent?(o)`

bogdan replied:
Maybe a comment is better here?

sgrif replied:
I'd prefer a method, which is less likely to get out of sync with the code.

bogdan replied:
done

sgrif commented:
This is definitely my preferred solution to the problem. I think we might need to have a deprecation cycle though, if there was previously a case where this could generate valid SQL, even if that SQL was incorrect.

bogdan commented:
@sgrif current regexp basically checks if there is a function call that has at least one comma inside like:

```
# match and raises
concat(field1, field2)
length('hello, world')
# not match and generates reverse
length(field1), length(field2)
```

Lets try to find a case when `,` inside `( )` can still result in valid SQL.

kaspth commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,17 @@
+*   `AR::Relation#reverse_order` throwns `AR::UnreversableOrderError` when the order can not be reversed
+    using current trival algorithm. Also raises the same errow when `#reverse_order` is called on
```
Typos in trivial and error here :wink:

kaspth commented on activerecord/lib/active_record/errors.rb:
```diff
@@ -241,4 +241,10 @@ class ImmutableRelation < ActiveRecordError
+
+  # UnreversableOrderError will be raised when #reverse_order can not reverse the order.
+  # ActiveRecord reverse order algorithm is not perfect and doesn't cover all possible use cases.
+  # That is why this exception can be raise when relation order is too complicated to be automatically reversed.
```
Some of your sentences retread the same territory. How about replacing these three lines with:

```ruby
# UnreversableOrderError is raised when a relation's order is too complex for
# +reverse_order+ to automatically reverse.
```

What do you think? :smile:

kaspth commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -1025,6 +1032,10 @@ def reverse_sql_order(order_query)
       end
     end
+    
+    def order_uses_function_with_multiple_arguments?(order)
```
Minor niggle, but can we switch function to method here?

sgrif replied:
function is more correct. Maybe rename to `uses_sql_function_with_multiple_arguments?` to make it more clear?

kaspth replied:
Ah, I missed that :+1:

--
Kasper

> Den 13/02/2015 kl. 18.42 skrev Sean Griffin <notifications@github.com>:
> 
> In activerecord/lib/active_record/relation/query_methods.rb:
> 
> > @@ -1025,6 +1032,10 @@ def reverse_sql_order(order_query)
> >          end
> >        end
> >      end
> > +    
> > +    def order_uses_function_with_multiple_arguments?(order)
> function is more correct. Maybe rename to uses_sql_function_with_multiple_arguments? to make it more clear?
> 
> —
> Reply to this email directly or view it on GitHub.
>

bogdan replied:
SQL is not object oriented that is why `function` is better

egilburg commented on activerecord/lib/active_record/errors.rb:
```diff
@@ -241,4 +241,10 @@ class ImmutableRelation < ActiveRecordError
   class TransactionIsolationError < ActiveRecordError
   end
+
+  # UnreversableOrderError will be raised when #reverse_order can not reverse the order.
```
I think it's spelled *Irreversible*

kaspth replied:
You're correct ;)

--
Kasper

> Den 13/02/2015 kl. 18.54 skrev Eugene Gilburg <notifications@github.com>:
> 
> In activerecord/lib/active_record/errors.rb:
> 
> > @@ -241,4 +241,10 @@ class ImmutableRelation < ActiveRecordError
> >    # The mysql, mysql2 and postgresql adapters support setting the transaction isolation level.
> >    class TransactionIsolationError < ActiveRecordError
> >    end
> > +
> > +  # UnreversableOrderError will be raised when #reverse_order can not reverse the order.
> I think it's spelled Irreversible
> 
> —
> Reply to this email directly or view it on GitHub.
>

bogdan commented:
* Renamed exception to `IrreversableOrder`
* Fixed typos in docs
* Renamed method

Thanks everyone for good comments.

vipulnsward commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,17 @@
+*   `AR::Relation#reverse_order` throwns `AR::IrreversableOrderError` when the order can not be reversed
```
`throws`

bogdan replied:
fixed

egilburg commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,17 @@
+*   `AR::Relation#reverse_order` throws `AR::IrreversableOrderError` when the order can not be reversed
```
sorry but this is still typo should be Irrevers**i**ble, not Irrevers**a**ble :-)

bogdan replied:
fixed. :+1:

egilburg commented on activerecord/test/cases/relations_test.rb:
```diff
@@ -188,6 +189,30 @@ def test_finding_with_reverted_assoc_order
+
+  def test_reverse_order_with_multiargument_function
+    assert_raises(ActiveRecord::IrreversibleOrderError) do
+      Topic.order("concat(author_name, title)").reverse_order
```
Why is a concat too complex to reverse? I understand the example with the nested DESC, but not this one. Unless I misread, this should still work if you swap the single ASC with DESC or vice versa, without changing anything in the function.

bogdan replied:
The following in stable rails: `Topic.order("concat(author_name, title)").reverse_order.to_sql` will give you `order by concat(author_name desc, title) desc`.

The goal of this patch is not to improve a way how rails reverse order but clearly mark cases that rails can not handle.

slbug replied:
Better to fix `reverse_order`imho

bogdan replied:
@slbug no matter how hard you try, you will not be able to support all use cases. This patch is not about making `reverse_order` better, but about confirming it is not perfect and will never be.

kaspth commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,17 @@
+*   `AR::Relation#reverse_order` throws `AR::IrreversibleOrderError` when the order can not be reversed
```
Let's spell out `ActiveRecord` here :smile:

bogdan replied:
Since we have no `ActiveResource` there is no doubt what is `AR`. Spelling it makes lines too long.

kaspth replied:
If it's so obvious what library these constants belong to, why even put AR there in the first place? :smile:

There's precedent in the file to either spell it out or leave it out. Go with one of those.

eileencodes replied:
I think we should spell it out. I don't think we should make the assumption it's obvious for developers who are new to Rails.

kaspth commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,17 @@
+          # After: raises AR::IrreversibleOrderError
+        Edge.all.reverse_order
+          # Before: SELECT `edges`.* FROM `edges` ORDER BY `edges`.`` DESC
+          # After: raises AR::IrreversibleOrderError
```
Here, and in the example just above, you have to spell out `ActiveRecord` as a matter of good documentation.

matthewd commented:
Is as much as this resembles https://github.com/rails/rails/pull/16779#issuecomment-54322316, I'm clearly :+1:.

But I suspect the current might-this-be-bad regexp is too narrow.

sgrif commented:
We should check for `NULLS FIRST` and friends as well.

bogdan commented:
@sgrif added `/nulls (first|last)\Z/i`. Something else?

bogdan commented:
@sgrif any time to come back to this issue?

sgrif commented:
It looks like you never pushed up the nulls first/last change. Can you do that and squash/rebase than ping me again?

bogdan committed:
Introduce ActiveRecord::IrreversibleOrderError

Raises when #reverse_order can not process SQL order instead of making
invalid SQL before this patch

bogdan commented:
@sgrif done.
